### ISSUE TITLE
Zone holding bitmap

### ISSUE THREAD

**Issue by blhoward2** [lurgh/infinitive-jkp717#1]
Hey there, I'm actually re-writing infinitive in node.js, while trying to improve some things. If you wouldn't mind helping me a bit. Are you seeing the zones holding bitmap actually change when you put a zone into holding on the thermostat? My system isn't zoned but that byte (11) in the 3b03 table is always 0 on mine. While I don't have a zoned system, I'm trying to write a complete driver.

Also, what do you see in byte 37 in 3b03 when you put a zone other than zone 1 into holding? That appears to be a 0/1 field showing whether zone 1 is in holding on mine, but I'm not sure what the point of that would be if other zones could also be holding... It just seems unnecessary given that there is a hold duration byte.

**Comment by lurgh** [comment]
Yes, byte 11 is 0x00 normally, 0x01 when zone 1 is in HOLD, 0x02 when zone 2 is in HOLD, etc.

Example: set zone 2 into HOLD, and then unset it:
*** [Dec 20 16:48:22] 2001 -> 9201: RESPONSE 003b030300000000030303030303**02**494900000000000052510000000000002d2d2d2d2d2d2d2d000000000000000000000000000000000000446f776e73746169727320005570737461697273202020005a4f4e4520332020202020005a4f4e4520342020202020005a4f4e4520352020202020005a4f4e4520362020202020005a4f4e4520372020202020005a4f4e452038202020202000
    [Dec 20 16:48:28] 2001 -> 9201: RESPONSE 003b030300000000030303030303**00**494900000000000052510000000000002d2d2d2d2d2d2d2d000000000000000000000000000000000000446f776e73746169727320005570737461697273202020005a4f4e4520332020202020005a4f4e4520342020202020005a4f4e4520352020202020005a4f4e4520362020202020005a4f4e4520372020202020005a4f4e452038202020202000

Regarding byte 37, what you're saying suggests that you may be confusing HOLD mode with a temporary setpoint override.  I had the same confusion at first because the tstat shows the word HOLD in both cases but it's actually a different state.  If you press the HOLD button on the tstat it will show "Now on HOLD" (with no timer) and you should see the per-zone bit flag set in byte 11.  If instead you change the setpoint, it will say "HOLD for 2:00" and this will set the per-zone bit flag in byte 37, and show the remaining number of minutes (0x0078 = 120) in the per-zone override time (bytes 38-39 for zone 1).  Does that make more sense?

**Comment by blhoward2** [comment]
Thanks, that helps a lot. You don't by chance have a real SAM do you?

**Comment by lurgh** [comment]
No I do not have a real SAM.  All my work has been based on the prior work that went into infinitive, plus my own experimentation.

**Comment by blhoward2** [comment]
Thanks for the help!

**Comment by lurgh** [comment]
hi @blhoward2 - just a heads-up in case you are curious, that i am rebasing my infinitive repo to @acd's fork, which evidently requires me to delete my original forked repo, so i am moving this issue around so it will eventually end up in my new fork rather than getting deleted with the old one.  i hope, anyway.

**Comment by blhoward2** [comment]
You should be able to rebase without deleting. Create a new branch. Force sync your master to the origin. Then merge in your copied branch. 

Or just do a regular rebase and force push: https://software.rcc.uchicago.edu/git/help/topics/git/git_rebase.md

**Comment by lurgh** [comment]
yes but i also want github to track mine as a fork of acd's rather than jkp717's fork of that, since acd is back to active development and it will make it easier stay in sync as we are both progressing.  are you aware of a way to accomplish that without deleting and re-forking?  or maybe i am making too much fuss about it?

**Comment by blhoward2** [comment]
I don't want to be presumptuous, but I'm fairly far along with replacing infinitive altogether. I just need to add mqtt and write an HA integration. It already has a Rest API. Mine is more robust than the current infinitive (a few new tables I decoded, logging, I'm working on the ability to emulate the thermostat for full testing ability) and I have some leads of extending it to touch thermostats too for a fully local solution for those too. (I just need to find a SAM or someone with one so I can see what the SAM sends to register itself with the touch.) Care to join me on mine? It's in Typescript. The integration I haven't started on will obviously be in Python.

**Comment by lurgh** [comment]
i can take a look and see how i could help out but my ts is pretty weak.  and i found that with mqtt support that i added, i could do everything i needed in HA (so far) without a custom python API or integration, which saved me a lot of time considering the old ones i found would need a lot of work and this looked like a single-user (just me) situation.

**Comment by blhoward2** [comment]
> yes but i also want github to track mine as a fork of acd's rather than jkp717's fork of that, since acd is back to active development and it will make it easier stay in sync as we are both progressing. are you aware of a way to accomplish that without deleting and re-forking? or maybe i am making too much fuss about it?

I don't think you can change the displayed origin of your work without opening a support request with GitHub. You can change the origin url on the command line and rebase off acd's without changing what it says it is a fork of, though. Have you talked to him at all? I'm very confused as to why he came from nowhere a month ago, made some fairly major changes, then disappeared again. Not sure I'd call that "back to active development."

I'm not sure if you use Home Assistant or not, but note that the current infinitive integration was made by someone else and will break fully from the recent acd changes. It isn't really maintained anymore and is already broken in a few ways.

**Comment by lurgh** [comment]
yes i've connected with acd and we discussed pulling my zoning and other work into his (now somewhat refactored) fork.  i then went dark for a few weeks but am picking it up again with some newfound free time.  i think we have the same observation of the existing HA integration and python API, which I am not using.

**Comment by blhoward2** [comment]
> i can take a look and see how i could help out but my ts is pretty weak. and i found that with mqtt support that i added, i could do everything i needed in HA (so far) without a custom python API or integration, which saved me a lot of time considering the old ones i found would need a lot of work and this looked like a single-user (just me) situation.

I'll add you to the repo (its currently private) and you can have a look. It's come a long way. I need to:
- Fix the zones bitmap (from this issue)
- Add support for setting temps and such (the api is currently readonly, but the all framework to encode and write packets exists I just need to add the functions and enable the api), 
- Add tests
- Add docs
- Do a little cleanup. 
 
It properly decodes packets (with error checking), refreshes the tables it needs, supports writing packets, etc. It fully supports all zones in the decoding and api, only displays enabled zones in the api, etc. And it's setup to be a driver so it can be easily extended for other things. (There are examples of some tooling for decoding things in utils/.)

Obviously if you're working with acd and prefer to keep working on it, the more the merrier. I wanted a ts project and don't know Go. And I was disappointed that everyone thus far seems to have stopped at the functionality they personally need. My goal is to write a comprehensive driver that supports up to the current touch thermostats. I've already bought -B and -C series Touch thermostats for testing, and confirmed with the author of infinitude that with a SAM he was able to enable the missing tables that were preventing Touch thermostats working before. I just need to convince him to capture the traffic or let me borrow his SAM so I can see what it sends when polled.

I'm also not sure how many users there are for this. I think there could be more with Touch support as that would bring it to modern installs. In any event, I wanted to test myself a bit with serial programming and such.

**Comment by lurgh** [comment]
I started with your parsers and it looks like you have covered a lot of ground.  There were a couple things I spotted that I might be able to add some perspective to on the zoning side, and a few good finds that I was looking for and knew had to be around there but never got down to.  Overall looks nice.

I'm not sure about getting too deep into coding but when ready I would be happy to try it out on my system and give you feedback and diagnostics.  As well as sharing what I discovered on my own about the zoning registers and calculations.

Also, as you say you plan to add mqtt, I would gladly collaborate both directions on that schema/API so that we can end up compatible.  As I mentioned, I find so far that mqtt is enough, and it lets me avoid a HA custom API/integration, websocket, etc.
